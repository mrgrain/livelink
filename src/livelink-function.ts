// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for LivelinkFunction
 */
export interface LivelinkFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/livelink.
 */
export class LivelinkFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: LivelinkFunctionProps) {
    super(scope, id, {
      description: 'src/livelink.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs16.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/livelink.lambda')),
    });
  }
}