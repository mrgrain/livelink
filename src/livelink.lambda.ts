import {
  APIGatewayProxyHandlerV2,
  APIGatewayProxyResultV2,
} from 'aws-lambda';
import fetch from 'node-fetch';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  try {
    console.log('INPUT', event.rawPath, event.rawQueryString);
    const stream = configure(event.rawPath);
    console.log('STREAM', stream);

    return await redirect(await resolveStream(stream));
  } catch {
    return { statusCode: 404 };
  }
};

function assert<T>(value: T | undefined | null): T {
  if (!value || value === null) {
    throw Error();
  }

  return value;
}

enum ContentType {
  DASH = 'video/vnd.mpeg.dash.mpd',
  HLS = 'application/vnd.apple.mpegurl',
}

interface Stream {
  url: string;
  contentType: ContentType;
  embed: boolean;
}

function configure(path: string): Stream {
  const match = assert(path.match(
    /^\/(channel|channel-id|video)(\/embed)?\/([^\/]*)(?=\.(m3u8|mpd))/,
  ));

  const url = (type: string, id: string): string => {
    switch (type.toLowerCase()) {
      case 'channel':
        return 'https://www.youtube.com/c/' + id + '/live';
      case 'channel-id':
        return 'https://www.youtube.com/channel/' +id + '/live';
      case 'video':
      default:
        return 'https://www.youtube.com/watch?v=' + id;
    }
  };

  return {
    url: url(assert(match[1]), assert(match[3])),
    contentType: assert(match[4]) === 'mpd' ? ContentType.DASH : ContentType.HLS,
    embed: Boolean(match[2]),
  };
}

async function resolveStream(
  stream: Stream,
): Promise<Stream> {
  const res = await fetch(stream.url);
  const body = await res.text();

  let matches;
  if (stream.contentType === ContentType.DASH) {
    matches = body.match(/(?<=dashManifestUrl":").*?(?=",")/g);
  } else {
    matches = body.match(/(?<=hlsManifestUrl":").*\.m3u8/g);
  }

  return {
    url: assert(matches?.[0]),
    contentType: stream.contentType,
    embed: stream.embed,
  };
}

async function redirect(
  stream: Stream,
): Promise<APIGatewayProxyResultV2> {
  if (stream.embed) {
    const response = await fetch(stream.url);
    return {
      statusCode: response.status,
      headers: {
        'Content-Type': stream.contentType,
      },
      body: await response.text(),
    };
  }

  return {
    statusCode: 302,
    headers: {
      'Content-Type': stream.contentType,
      'location': stream.url,
    },
  };
}
