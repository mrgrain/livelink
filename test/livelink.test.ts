import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { LivelinkFunction } from '../src';

describe('Livelink', () => {
  test('creates a Function', () => {
    const stack = new cdk.Stack();
    new LivelinkFunction(stack, 'MyLivelink');

    const template = Template.fromStack(stack);
    expect(template).toMatchSnapshot();
  });
});