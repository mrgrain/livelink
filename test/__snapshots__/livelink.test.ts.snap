// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Livelink creates a Function 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "MyLivelink43DFC310": Object {
      "DependsOn": Array [
        "MyLivelinkServiceRole1E5C63C4",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "53fc90a81431688ece6c4f953aeed7d8d9effc01c427ef8f53369eac3cf2dafc.zip",
        },
        "Description": "src/livelink.lambda.ts",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyLivelinkServiceRole1E5C63C4",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyLivelinkServiceRole1E5C63C4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
